desc/compute/_stability.py

```diff
@@ -232,3 +232,399 @@ def _magnetic_well(params, transforms, profiles, data, **kwargs):
+
+    zeta0 = kwargs.get("zeta0", jnp.linspace(-0.5 * jnp.pi, 0.5 * jnp.pi, 15))
+    N_zeta0 = len(zeta0)
+
+    # This would fail with rho vectorization
+    iota = jnp.mean(data["iota"])
+    shear = jnp.mean(data["shear"])
+    psi = jnp.mean(data["psi"])
+    sign_psi = jnp.sign(psi)
+    sign_iota = jnp.sign(iota)
+
+    N_rho = int(source_grid.num_rho)
+    N_alpha = int(source_grid.num_alpha)
+
+    # phi is the same for each alpha
+    phi = source_grid.nodes[:: N_rho * N_alpha, 2]
+    N_zeta = len(phi)
+
+    B = source_grid.meshgrid_reshape(data["|B|"], "arz")
+    B_sup_zeta = source_grid.meshgrid_reshape(data["B^zeta"], "arz")
+    gradpar = B_sup_zeta / B
+
+    # This would fail with rho vectorization
+    dpdpsi = jnp.mean(mu_0 * data["p_r"] / data["psi_r"])
+
+    g_sup_aa = source_grid.meshgrid_reshape(data["g^aa"], "arz")[None, ...]
+    g_sup_ra = source_grid.meshgrid_reshape(data["g^ra"], "arz")[None, ...]
+    g_sup_rr = source_grid.meshgrid_reshape(data["g^rr"], "arz")[None, ...]
+
+    gds2 = jnp.reshape(
+        rho**2
+        * (
+            g_sup_aa
+            - 2 * sign_iota * shear / rho * zeta0[:, None, None, None] * g_sup_ra
+            + zeta0[:, None, None, None] ** 2 * (shear / rho) ** 2 * g_sup_rr
+        ),
+        (N_alpha, N_zeta0, N_zeta),
+    )
+
+    f = a_N * B_N**3 * gds2 / B**3 * 1 / gradpar
+    g = a_N**3 * B_N * gds2 / B * gradpar
+    g_half = (g[:, :, 1:] + g[:, :, :-1]) / 2
+
+    cvdrift = source_grid.meshgrid_reshape(data["cvdrift"], "arz")[None, ...]
+    cvdrift0 = source_grid.meshgrid_reshape(data["cvdrift0"], "arz")[None, ...]
+
+    c = (
+        a_N**3
+        * B_N
+        * jnp.reshape(
+            2
+            / B_sup_zeta[None, ...]
+            * sign_psi
+            * rho**2
+            * dpdpsi
+            * (cvdrift - shear / (2 * rho**2) * zeta0[:, None, None, None] * cvdrift0),
+            (N_alpha, N_zeta0, N_zeta),
+        )
+    )
+
+    h = phi[1] - phi[0]
+
+    i = jnp.arange(N_alpha)[:, None, None, None]
+    l = jnp.arange(N_zeta0)[None, :, None, None]
+    j = jnp.arange(N_zeta - 2)[None, None, :, None]
+    k = jnp.arange(N_zeta - 2)[None, None, None, :]
+
+    A = jnp.zeros((N_alpha, N_zeta0, N_zeta - 2, N_zeta - 2))
+    B_inv = jnp.zeros((N_alpha, N_zeta0, N_zeta - 2, N_zeta - 2))
+
+    A = A.at[i, l, j, k].set(
+        g_half[i, l, k] / h**2 * (j - k == -1)
+        + (-(g_half[i, l, j + 1] + g_half[i, l, j]) / h**2 + c[i, l, j + 1])
+        * (j - k == 0)
+        + g_half[i, l, j] / h**2 * (j - k == 1)
+    )
+
+    B_inv = B_inv.at[i, l, j, k].set(1 / jnp.sqrt(f[i, l, j + 1]) * (j - k == 0))
+
+    A_redo = B_inv @ A @ jnp.transpose(B_inv, axes=(0, 1, 3, 2))
+
+    w, _ = jnp.linalg.eigh(A_redo)
+    # max over "zeta" axis, still a function of rho, alpha, zeta0
+    gamma = jnp.real(jnp.max(w, axis=(2,)))
+
+    data["ideal ball lambda"] = gamma.flatten()
+
+    return data
+
+
+@register_compute_fun(
+    name="Newcomb ball metric",
```

Comment:
Similar to previous comment, I don't think we should use "ball" as an abbreviation for "ballooning" since "ball" is itself a real word that has other meanings